### PR TITLE
Implement mixin generation 

### PR DESCRIPTION
**Pull request by jskeet**
This won't build on CI as Google.Cloud.Location isn't released yet.

Various things still to do, mostly:

- Check whether routing header customization is an issue
- Generate proper XML docs instead of FIXME
- Work out whether the default mixin settings are appropriate

However, the result of generation does genuinely work:
https://github.com/googleapis/google-cloud-dotnet/pull/7426

(I've just tested listing locations, but that worked first time.)

### CODE REVIEW

File: Google.Api.Generator/CodeGenerator.cs [hunk 1]
```diff
@@ -142,7 +166,7 @@ internal static class CodeGenerator
                 foreach (var service in fileDesc.Services)
                 {
                     // Generate settings and client code for requested package.
-                    var serviceDetails = new ServiceDetails(catalog, ns, service, grpcServiceConfig);
+                    var serviceDetails = new ServiceDetails(catalog, ns, service, grpcServiceConfig, serviceConfig);
                     allServiceDetails.Add(serviceDetails);
 
                     var ctx = SourceFileContext.CreateFullyAliased(clock, s_wellknownNamespaceAliases);
```
Comment (amanda-tarafa) [inline Google.Api.Generator/CodeGenerator.cs:172]: Leaving this comment here as I can't write further down.

I think we should keep track of which mixinx we have and add them to the project file we generate (similar to what we do for LRO). Even if we ourselves discard those, we should produce complete project files I believe.

**Review by amanda-tarafa — COMMENTED** [review]
Looks really good. I've left a couple of comments.

File: Google.Api.Generator/CodeGenerator.cs [hunk 2]
```diff
@@ -84,19 +85,22 @@ internal static class CodeGenerator
...
         public static IEnumerable<ResultFile> Generate(IReadOnlyList<FileDescriptorProto> descriptorProtos, IEnumerable<string> filesToGenerate, IClock clock,
-            string grpcServiceConfigPath, IEnumerable<string> commonResourcesConfigPaths)
+            string grpcServiceConfigPath, string serviceConfigYamlPath, IEnumerable<string> commonResourcesConfigPaths)
         {
             var descriptors = FileDescriptor.BuildFromByteStrings(descriptorProtos.Select(proto => proto.ToByteString()), s_registry);
             // Load side-loaded configurations; both optional.
-            var grpcServiceConfig = grpcServiceConfigPath != null ? ServiceConfig.Parser.ParseJson(File.ReadAllText(grpcServiceConfigPath)) : null;
+            var grpcServiceConfig = grpcServiceConfigPath is object ? ServiceConfig.Parser.ParseJson(File.ReadAllText(grpcServiceConfigPath)) : null;
+            var serviceConfig = ParseServiceConfigYaml(serviceConfigYamlPath);
+            File.WriteAllText("c:/users/jonskeet/test/config.json", serviceConfig.ToString());
```
[the remark this answers is not in the document]
Comment (jskeet) [reply Google.Api.Generator/CodeGenerator.cs]: Humbug - I could have sworn I already had done (having spotted it when I was about to create the PR)! Pushing now...

File: Google.Api.Generator/CodeGenerator.cs [hunk 3]
```diff
@@ -116,15 +120,35 @@ internal static class CodeGenerator
...
+            
+            var deserializer = new Deserializer();
+            using (var reader = File.OpenText(path))
+            {
+                var yamlObject = deserializer.Deserialize(reader);
+                var serializer = new SerializerBuilder().JsonCompatible().Build();
+                var writer = new StringWriter();
+                serializer.Serialize(writer, yamlObject);
+                string json = writer.ToString();
+                var parser = new JsonParser(JsonParser.Settings.Default.WithIgnoreUnknownFields(true));
```
[the remark this answers is not in the document]
Comment (jskeet) [reply Google.Api.Generator/CodeGenerator.cs:144]: Possibly - but it would be significantly more work. I don't find the YAML library terribly easy to use, whereas just converting to JSON and then using that JSON is much simpler :)

File: Google.Api.Generator/CodeGenerator.cs (hunk 1 above)
Comment (jskeet) [reply Google.Api.Generator/CodeGenerator.cs:171]: Ooh yes, good plan. Will do that next week.

File: Google.Api.Generator/Generation/ServiceDetails.cs [hunk 4]
```diff
@@ -62,6 +70,7 @@ public ServiceDetails(ProtoCatalog catalog, string ns, ServiceDescriptor desc, S
             SnippetsClientName = $"{desc.Name.ToLowerCamelCase()}Client";
             UnitTestsTyp = Typ.Manual(UnitTestsNamespace, $"Generated{desc.Name}ClientTest");
             NonStandardLro = NonStandardLroDetails.ForService(desc);
+            Mixins = serviceConfig?.Apis.Select(api => AvailableMixins.GetValueOrDefault(api.Name)).Where(mixin => mixin is object).ToList() ?? Enumerable.Empty<MixinDetails>();
```
[the remark this answers is not in the document]
Comment (jskeet) [reply Google.Api.Generator/Generation/ServiceDetails.cs:73]: I'm not sure - a service yaml file can contain multiple services which aren't all mixins. Worth considering though - extraneous ones may be stripped anyway.

File: Google.Api.Generator/CodeGenerator.cs [hunk 5]
```diff
@@ -84,19 +85,21 @@ internal static class CodeGenerator
         };
 
         public static IEnumerable<ResultFile> Generate(FileDescriptorSet descriptorSet, string package, IClock clock,
-            string grpcServiceConfigPath, IEnumerable<string> commonResourcesConfigPaths)
+            string grpcServiceConfigPath, string serviceConfigYamlPath, IEnumerable<string> commonResourcesConfigPaths)
```
Comment (viacheslav-rostovtsev) [inline Google.Api.Generator/CodeGenerator.cs:88]: just 'serviceConfigPath' for consistency with tests?

File: Google.Api.Generator/CodeGenerator.cs [hunk 6]
```diff
@@ -84,19 +85,21 @@ internal static class CodeGenerator
...
         }
 
         public static IEnumerable<ResultFile> Generate(IReadOnlyList<FileDescriptorProto> descriptorProtos, IEnumerable<string> filesToGenerate, IClock clock,
-            string grpcServiceConfigPath, IEnumerable<string> commonResourcesConfigPaths)
+            string grpcServiceConfigPath, string serviceConfigYamlPath, IEnumerable<string> commonResourcesConfigPaths)
         {
             var descriptors = FileDescriptor.BuildFromByteStrings(descriptorProtos.Select(proto => proto.ToByteString()), s_registry);
             // Load side-loaded configurations; both optional.
-            var grpcServiceConfig = grpcServiceConfigPath != null ? ServiceConfig.Parser.ParseJson(File.ReadAllText(grpcServiceConfigPath)) : null;
+            var grpcServiceConfig = grpcServiceConfigPath is object ? ServiceConfig.Parser.ParseJson(File.ReadAllText(grpcServiceConfigPath)) : null;
```
Comment (viacheslav-rostovtsev) [inline Google.Api.Generator/CodeGenerator.cs:100]: nit: changing `!= null` to `is object` makes me sad

File: Google.Api.Generator/CodeGenerator.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (viacheslav-rostovtsev) [reply Google.Api.Generator/CodeGenerator.cs:144]: I ended up going through json in my attempt as well, altho in my case I went for ` Service.Parser.ParseJson()`.

File: Google.Api.Generator/CodeGenerator.cs (hunk 6 above)
Comment (jskeet) [reply Google.Api.Generator/CodeGenerator.cs:100]: It's the idiomatic approach these days.

File: Google.Api.Generator/CodeGenerator.cs (hunk 5 above)
Comment (jskeet) [reply Google.Api.Generator/CodeGenerator.cs:88]: *Could* do, although given that we've had the gRPC service config this whole time, it feels worth clarifying which one we're really talking about.

**Comment by jskeet** [comment]
Okay, I've revisited this now:

- We now have XML docs
- We don't yet have any tests (needs addressing)
- We don't yet include mix-in dependences in the project file (would be nice, but could be deferred)
- We generate the mix-ins for *every* client - is that genuinely what we want? (General design question really.)

@viacheslav-rostovtsev: Do you think it's best to add a new ProtoTests directory for Mixins, that includes both locations and IAM?

File: Google.Api.Generator/CodeGenerator.cs (hunk 1 above)
Comment (jskeet) [reply Google.Api.Generator/CodeGenerator.cs:182]: This is now done.

File: Google.Api.Generator/CodeGenerator.cs (hunk 5 above)
Comment (jskeet) [reply Google.Api.Generator/CodeGenerator.cs:88]: Done in the end.

**Comment by jskeet** [comment]
This is now ready for review. I've separated it into two commits - one for the implementation and one for the new tests.

File: Google.Api.Generator/Generation/ServiceAbstractClientClassCodeGenerator.cs [hunk 7]
```diff
@@ -60,9 +60,11 @@ private ClassDeclarationSyntax Generate()
                 var createFromCallInvoker = CreateFromCallInvoker();
                 var shutdown = ShutdownDefaultChannelsAsync(channelPool, create, createAsync);
                 var grpcClient = GrpcClient();
+                var mixins = Mixins();
```
Comment (viacheslav-rostovtsev) [inline Google.Api.Generator/Generation/ServiceAbstractClientClassCodeGenerator.cs:63]: Nit: move this closer to usage or inline?

File: Google.Api.Generator/Generation/ServiceDetails.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (viacheslav-rostovtsev) [reply Google.Api.Generator/Generation/ServiceDetails.cs:81]: If something is added in service yaml that is not a known mixin and we don't have it in the request to generate, it might be e.g. a misspelled mixin (`Locattions`).
Making a check here and failing might be a good idea.

**Review by viacheslav-rostovtsev — APPROVED** [review]

**Review by jskeet — COMMENTED** [review]
Only significant change is in the new last commit.

**Review by amanda-tarafa — COMMENTED** [review]
The mixins LGTM.
But there's a change from the snippet metadata PR that you are inandvertently reverting I think.

And one comment and one question on mixins:

- You asked if mixins should be generated for every client. My two cents is
  - Yes, as we don't have the information of which resources can the mixins be used on. (If we had that, we could restrict the mixins to services that handle those resources only, and we could even add client methods to apply the mixins to the relevant resources only, or helper methods that receive thet accepted resource names only, etc.).
  - Or we generate the mixins for none of the clients and just generate some other general way of accessing the clients for the mixins.
  Basically, generating the mixins for some of the clients but not for others seems random as we don't have enough config information.
- Should we generate some snippets for mixins? That can all be done later of course.

**Comment by jskeet** [comment]
Thanks for the review. I *thought* I'd resolved things properly, but I may have missed something - will check, thanks.
(I'll do that and reply to individual comments tomorrow.)

File: Google.Api.Generator/Generation/ServiceCodeGenerator.cs [hunk 8]
```diff
@@ -219,5 +220,37 @@ private static IEnumerable<MemberDeclarationSyntax> LroPartialClasses(SourceFile
...
+                    var grpcInnerClass = Class(Public | Partial, svc.GrpcClientTyp);
+                    using (ctx.InClass(grpcInnerClass))
+                    {
+                        var callInvoker = Property(Private, ctx.TypeDontCare, "CallInvoker");
+
+                        foreach (var mixin in svc.Mixins)
+                        {
+                            var grpcClientType = ctx.Type(mixin.GrpcClientType);
+                            var createClientMethod = Method(Public | Virtual, grpcClientType, "Create" + mixin.GrpcClientType.Name)()
+                                .WithBody(New(grpcClientType)(callInvoker))
```
[the remark this answers is not in the document]
Comment (jskeet) [reply Google.Api.Generator/Generation/ServiceCodeGenerator.cs:242]: I don't believe so. If it does, we can cross that bridge when we come to it.

File: Google.Api.Generator/Generation/CsProjGenerator.cs [hunk 9]
```diff
@@ -23,14 +25,21 @@ internal static class CsProjGenerator
         private const string GaxGrpcCoreVersion = "[3.6.0, 4.0.0)";
         private const string GrpcCoreVersion = "[2.41.0, 3.0.0)";
         private const string LroVersion = "[2.3.0, 3.0.0)";
+        private const string IamVersion = "[2.3.0, 3.0.0)";
+        private const string LocationVersion = "[1.0.0-beta01, 2.0.0)";
```
[the remark this answers is not in the document]
Comment (jskeet) [reply Google.Api.Generator/Generation/CsProjGenerator.cs:29]: Will release a Location 1.0 and update the PR to use that.

File: Google.Api.Generator/CodeGenerator.cs [hunk 10]
```diff
@@ -159,8 +183,8 @@ internal static class CodeGenerator
                 foreach (var service in fileDesc.Services)
                 {
                     // Generate settings and client code for requested package.
-                    var serviceDetails = new ServiceDetails(catalog, ns, service, grpcServiceConfig);
-                    packageServiceDetails.Add(serviceDetails);
+                    var serviceDetails = new ServiceDetails(catalog, ns, service, grpcServiceConfig, serviceConfig);
+                    allServiceDetails.Add(serviceDetails);
```
[the remark this answers is not in the document]
Comment (jskeet) [reply Google.Api.Generator/CodeGenerator.cs:214]: I've definitely rebased. Not sure how the tests were passing before. Have reverted that change... will have a closer look after repushing.

File: Google.Api.Generator/Generation/ServiceDetails.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jskeet) [reply Google.Api.Generator/Generation/ServiceDetails.cs:81]: (This is now addressed in an additional commit.)

**Review by amanda-tarafa — APPROVED** [review]
LGTM
